### PR TITLE
Add support for thirdparty resource watch/create/put and delete operations

### LINKED ISSUE #201: Support third party resources of a given Kind
Hi,

Thanks for this! This does exactly what I need. I am however very lost on how to handle the following use case

I've created a new thirdparty resource of eg. type `database`

How do I watch the `/apis/database.k8s.com/v1/databases` endpoint to get events when a new resource of type database is created or deleted.

I understand I'd have to sub-class one of the classes to create an endpoint? however I'm not sure which one?

Would appreciate any help with this.

Thanks.

### PR DESCRIPTION
**Pull request by jonathan-kosgei**
Fixes #201

```release-note
Support third-party resources
```

### CODE REVIEW

**Comment by jonathan-kosgei** [comment]
How do I name the classes in swagger? So that we don't get something like DefaultAPi as the api_instance object?

**Comment by mbohlool** [comment]
@jonathan-kosgei add a tag to operations. Look at our swagger.json for examples.

File: scripts/preprocess_spec.py [hunk 1]
```diff
@@ -102,7 +102,17 @@ def strip_tags_from_operation_id(operation, _):
     operation['operationId'] = operation_id
 
 
+def add_thirdparty_resource_paths(spec):
+    with open('thirdpartypaths.json', 'r') as third_party_spec_file:
+        third_party_spec = json.loads(third_party_spec_file.read())
+    for path in third_party_spec.keys():
+        spec['paths'][path] = third_party_spec[path]
```
Comment (mbohlool) [inline scripts/preprocess_spec.py]: Can you just add a validation here to make sure we are not overwriting an existing path?

**Comment by mbohlool** [comment]
one nit comment. Let me know when you added the tags for better naming and one example. This looks good so far.

(btw, use `git commit --amend` for small commits that should be part of the previous commit. you can merge that one small commit by `git rebase -i HEAD~2`)

**Comment by jonathan-kosgei** [comment]
Awesome tip. Thanks. Pushing changes, including a example. I'll add a few more as I go along.

File: examples/create_thirdparty_resource.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+from __future__ import print_function
+import time
+import kube_resource
+import json
+from kube_resource.rest import ApiException
+from pprint import pprint
+
+with open('/var/run/secrets/kubernetes.io/serviceaccount/token', 'r') as token_file:
```
Comment (mbohlool) [inline examples/create_thirdparty_resource.py]: Please use load_kube_config like other examples.

File: examples/create_thirdparty_resource.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+kube_resource.configuration.api_key['Authorization'] = token
+kube_resource.configuration.api_key_prefix['Authorization'] = 'Bearer'
+kube_resource.configuration.ssl_ca_cert = '/var/run/secrets/kubernetes.io/serviceaccount/ca.crt'
+api_instance = kube_resource.DefaultApi()
+
+namespace = 'default'
+resource = 'repos'
+fqdn = 'git.k8s.com'
+
+body = json.loads('{  "apiVersion": "git.k8s.com/v1",  "kind": "RePo",  "metadata": {    "name": "blog-repo"  },  "repo": "github.com/user/my-blog",  "oauth": 123456789,  "branch": "master",  "key": "user-git-deploy-secret",  "path": "/path/in/volume",  "image": "image to run job in",  "then": "hugo --destination=/home/user/hugosite/public",  "pvc": "persistent-volume-claim",  "username": "repo username",  "password": "repo password"}')
```
Comment (mbohlool) [inline examples/create_thirdparty_resource.py]: I would just make it a python map. The syntax is almost identical to json.

File: examples/create_thirdparty_resource.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+from kube_resource.rest import ApiException
+from pprint import pprint
+
+with open('/var/run/secrets/kubernetes.io/serviceaccount/token', 'r') as token_file:
+    token=token_file.read()
+
+kube_resource.configuration.api_key['Authorization'] = token
+kube_resource.configuration.api_key_prefix['Authorization'] = 'Bearer'
+kube_resource.configuration.ssl_ca_cert = '/var/run/secrets/kubernetes.io/serviceaccount/ca.crt'
+api_instance = kube_resource.DefaultApi()
```
Comment (mbohlool) [inline examples/create_thirdparty_resource.py]: I thought you renamed DefaultAPi using tags?

File: scripts/thirdpartypaths.json [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+        {
+          "Bearer": [
+
+          ]
+        }
+      ],
+      "summary": "Gets Resources",
+      "description": "Returns a list of Resources",
+      "tags": [
+        "thirdparty_v1"
```
Comment (mbohlool) [inline scripts/thirdpartypaths.json]: Name Suggestion : third_party_resource?

Comment (mbohlool) [inline scripts/thirdpartypaths.json]: third_party_resources

**Comment by mbohlool** [comment]
Look like there are some formatting error too. Use script/update-pep8.sh

**Comment by jonathan-kosgei** [comment]
On it.

**Comment by pvdvreede** [comment]
This is something I would really like so that I can use the python client to create a custom operator. Will this be merged/released soon? 

Is there a current way to watch a specific third party resource I create in version 2.0 as a stop gap?

**Comment by mbohlool** [comment]
@jonathan-kosgei any update on this? Can I help with anything?
I too hope this merge in soon.

**Comment by mbohlool** [comment]
The code looks good. I will Squash and merge after tests passed.

**Comment by jonathan-kosgei** [comment]
Hi @pvdvreede Sorry, got caught up with work, fixed

**Comment by pvdvreede** [comment]
Great! Thanks very much! Hopefully a new release can be done soon?

**Comment by mbohlool** [comment]
@jonathan-kosgei The client needs to be updated with the new yaml changes. I will send a follow up PR tomorrow for that.
@pvdvreede We need to finalize the interface and add some more tests before we cut a release. Would be nice if you can install the client from the master and give this a try and maybe add some examples so we know the interface is good enough. I will do some work on this soon.

**Comment by pvdvreede** [comment]
Thanks @mbohlool I gave this a go last night and used the master branch. It errored saying there was no `ThirdPartyResources` attribute on kubernetes module. So I am not sure how to get the api instance. Does the example need updating/changing?

**Comment by jonathan-kosgei** [comment]
I think probably the client needs to be regenerated? 
You can clone the code locally and run the update-client script in the scripts folder. In the meantime take a look at https://github.com/jonathan-kosgei/kubeResource, I've tested that and used it as the base for this pr

**Pull request merged**